### PR TITLE
use visitor for #[structural_match] check

### LINKED ISSUE #61188: ICE when trying to match on non-PartialEq slice.
The following code will result in an ICE in ~~codegen~~ MIR construction:

```rust
struct B(i32);

const A: &[B] = &[];

fn main() {
    match &[][..] {
        A => (),
        _ => (),
    }
}
```

[playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=4db539e3f11d582083944d361b64cec9)

<details>

<summary>Error and Backtrace</summary>

```text
error: internal compiler error: src/librustc/traits/codegen/mod.rs:156: Encountered errors `[FulfillmentError(Obligation(predicate=Binder(TraitPredicate(<B as std::cmp::PartialEq>)),depth=2),Unimplemented)]` resolving bounds after type-checking

thread 'rustc' panicked at 'Box<Any>', src/librustc_errors/lib.rs:570:9
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
stack backtrace:

 0: std::sys::unix::backtrace::tracing::imp::unwind_backtrace
             at src/libstd/sys/unix/backtrace/tracing/gcc_s.rs:39
   1: std::sys_common::backtrace::_print
             at src/libstd/sys_common/backtrace.rs:71
   2: std::panicking::default_hook::{{closure}}
             at src/libstd/sys_common/backtrace.rs:59
             at src/libstd/panicking.rs:197
   3: std::panicking::default_hook
             at src/libstd/panicking.rs:211
   4: rustc::util::common::panic_hook
   5: std::panicking::rust_panic_with_hook
             at src/libstd/panicking.rs:478
   6: std::panicking::begin_panic
   7: rustc_errors::Handler::span_bug
   8: rustc::util::bug::opt_span_bug_fmt::{{closure}}
   9: rustc::ty::context::tls::with_opt::{{closure}}
  10: rustc::ty::context::tls::with_context_opt
  11: rustc::ty::context::tls::with_opt
  12: rustc::util::bug::opt_span_bug_fmt
  13: rustc::util::bug::span_bug_fmt
  14: rustc::traits::codegen::<impl rustc::infer::InferCtxt>::drain_fulfillment_cx_or_panic
  15: rustc::ty::context::GlobalCtxt::enter_local
  16: rustc::traits::codegen::codegen_fulfill_obligation
  17: rustc::ty::query::__query_compute::codegen_fulfill_obligation
  18: rustc::ty::query::<impl rustc::ty::query::config::QueryAccessors for rustc::ty::query::queries::codegen_fulfill_obligation>::compute
  19: rustc::dep_graph::graph::DepGraph::with_task_impl
  20: rustc::ty::query::plumbing::<impl rustc::ty::context::TyCtxt>::get_query
  21: rustc::ty::instance::Instance::resolve
  22: rustc_mir::lints::check
  23: rustc::ty::context::GlobalCtxt::enter_local
  24: rustc_mir::build::mir_build
  25: rustc_mir::transform::mir_built
  26: rustc::ty::query::__query_compute::mir_built
  27: rustc::ty::query::<impl rustc::ty::query::config::QueryAccessors for rustc::ty::query::queries::mir_built>::compute
  28: rustc::dep_graph::graph::DepGraph::with_task_impl
  29: rustc::ty::query::plumbing::<impl rustc::ty::context::TyCtxt>::get_query
  30: rustc_mir::transform::check_unsafety::unsafety_check_result
  31: rustc::ty::query::__query_compute::unsafety_check_result
  32: rustc::ty::query::<impl rustc::ty::query::config::QueryAccessors for rustc::ty::query::queries::unsafety_check_result>::compute
  33: rustc::dep_graph::graph::DepGraph::with_task_impl
  34: rustc::ty::query::plumbing::<impl rustc::ty::context::TyCtxt>::get_query
  35: rustc_mir::transform::mir_const
  36: rustc::ty::query::__query_compute::mir_const
  37: rustc::ty::query::<impl rustc::ty::query::config::QueryAccessors for rustc::ty::query::queries::mir_const>::compute
  38: rustc::dep_graph::graph::DepGraph::with_task_impl
  39: rustc::ty::query::plumbing::<impl rustc::ty::context::TyCtxt>::get_query
  40: rustc_mir::transform::mir_validated
  41: rustc::ty::query::__query_compute::mir_validated
  42: rustc::ty::query::<impl rustc::ty::query::config::QueryAccessors for rustc::ty::query::queries::mir_validated>::compute
  43: rustc::dep_graph::graph::DepGraph::with_task_impl
  44: rustc::ty::query::plumbing::<impl rustc::ty::context::TyCtxt>::get_query
  45: rustc_mir::borrow_check::mir_borrowck
  46: rustc::ty::query::__query_compute::mir_borrowck
  47: rustc::ty::query::<impl rustc::ty::query::config::QueryAccessors for rustc::ty::query::queries::mir_borrowck>::compute
  48: rustc::dep_graph::graph::DepGraph::with_task_impl
  49: rustc::ty::query::plumbing::<impl rustc::ty::context::TyCtxt>::get_query
  50: rustc::ty::<impl rustc::ty::context::TyCtxt>::par_body_owners
  51: rustc::util::common::time
  52: rustc_interface::passes::analysis
  53: rustc::ty::query::__query_compute::analysis
  54: rustc::ty::query::<impl rustc::ty::query::config::QueryAccessors for rustc::ty::query::queries::analysis>::compute
  55: rustc::dep_graph::graph::DepGraph::with_task_impl
  56: rustc::ty::query::plumbing::<impl rustc::ty::context::TyCtxt>::get_query
  57: rustc_interface::passes::BoxedGlobalCtxt::access::{{closure}}
  58: rustc_interface::passes::create_global_ctxt::{{closure}}
  59: rustc_interface::interface::run_compiler_in_existing_thread_pool
  60: std::thread::local::LocalKey<T>::with
  61: scoped_tls::ScopedKey<T>::set
  62: syntax::with_globals
query stack during panic:
#0 [codegen_fulfill_obligation] checking if `std::cmp::PartialEq` fulfills its obligations
#1 [mir_built] processing `main`
#2 [unsafety_check_result] processing `main`
#3 [mir_const] processing `main`
#4 [mir_validated] processing `main`
#5 [mir_borrowck] processing `main`
#6 [analysis] running analysis passes on this crate
end of query stack
```

</details>

### LINKED ISSUE #62307: structural_match bug:  deep refs like `& &B` leak use of PartialEq
(Spawned off of investigation of #61188)

rust-lang/rfcs#1445 says that we're supposed to reject uses of consts with ADT's that don't implement `#[structural_match]` (i.e. ADT's that do not do `#[derive(PartialEq, Eq)]`.

Unfortunately, we did not quite finish the job. Consider the following code ([play](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=92cbbf1bf22e561eb05ee48a902a94eb)):

```rust
#[derive(Debug)]
struct B(i32);

// Overriding PartialEq to use this strange notion of equality exposes
// whether `match` is using structural-equality or method-dispatch
// under the hood, which is the antithesis of rust-lang/rfcs#1445
impl PartialEq for B {
    fn eq(&self, other: &B) -> bool { std::cmp::min(self.0, other.0) == 0 }
}

const ZER_B: & & B = & & B(0);
const ONE_B: & & B = & & B(1);

fn main() {
    if let ONE_B = ZER_B {
        println!("CLAIM Z: {:?} matches {:?}", ONE_B, ZER_B);
    }

    if let ONE_B = ONE_B {
        println!("CLAIM O: {:?} matches {:?}", ONE_B, ONE_B);
    }
}
```

This prints:

```text
CLAIM Z: B(1) matches B(0)
```

but I would expect to see something like the error you get if you use constants of type `&B` instead of `& &B`:

```text
error: to use a constant of type `B` in a pattern, `B` must be annotated with `#[derive(PartialEq, Eq)]`
  --> src/main.rs:15:12
   |
15 |     if let ONE_B = ZER_B {
   |            ^^^^^
```

### PR DESCRIPTION
**Pull request by pnkfelix**
This changes the code so that we recur down the structure of a type of a const (rather than just inspecting at a shallow one or two levels) when we are looking to see if it has an ADT that did not derive `PartialEq` and `Eq`.

Fix #61188 

Fix #62307 

Cc #62336

### CODE REVIEW

**Comment by pnkfelix** [comment]
The way this is written might be bad for performance, since it is doing a visit of the type structure on every call to `const_to_pat`, and that routine calls itself recursively.

Thinking further on this, we *might* be able to avoid that cost by factoring the recursive traversal into a subroutine, and then doing the type traversal once, at the start of `const_to_pat`, before starting the recursive traversal. (But I would want to be careful about not overlooking cases where the problematic ADT occurs more deeply in the input.)

**Comment by pnkfelix** [comment]
(I've left the WIP marker because I want to add tests to double-check that the refactoring in commit 9db458f300a1f0c844071a41a547bd6294c99610 didn't break things.)

File: src/librustc_mir/hair/pattern/mod.rs [hunk 1]
```diff
@@ -1058,6 +1067,52 @@ impl<'a, 'tcx> PatternContext<'a, 'tcx> {
     }
 }
 
+fn search_for_adt_without_structural_match<'tcx>(tcx: TyCtxt<'tcx>,
```
Comment (Centril) [inline src/librustc_mir/hair/pattern/mod.rs]: Indentation is weird here, should be:
```rust
fn search_for_adt_without_structural_match<'tcx>(
    tcx: TyCtxt<'tcx>,
    ty: Ty<'tcx>,
) -> Option<&'tcx AdtDef> {
    ...
```

File: src/librustc_mir/hair/pattern/mod.rs [hunk 2]
```diff
@@ -1058,6 +1067,52 @@ impl<'a, 'tcx> PatternContext<'a, 'tcx> {
...
+            if let ty::Adt(adt_def, _) = ty.sty {
+                if !self.tcx.has_attr(adt_def.did, sym::structural_match) {
+                    self.found = Some(&adt_def);
+                    debug!("SearchForAdtWithoutStructuralMatch found adt_def: {:?}", adt_def);
+                    return true // Halt visiting
+                }
+
+                // For ADT, we do *not* visit contents.
+                //
+                // The presence of the #[structural_match] attribute
```
Comment (Centril) [inline src/librustc_mir/hair/pattern/mod.rs]:
```suggestion
                // The presence of the `#[structural_match]` attribute
```

File: src/librustc_mir/hair/pattern/mod.rs [hunk 3]
```diff
@@ -1058,6 +1067,52 @@ impl<'a, 'tcx> PatternContext<'a, 'tcx> {
...
+    struct Search<'tcx> { tcx: TyCtxt<'tcx>, found: Option<&'tcx AdtDef> }
+    impl<'tcx> TypeVisitor<'tcx> for Search<'tcx> {
+        fn visit_ty(&mut self, ty: Ty<'tcx>) -> bool {
+            debug!("SearchForAdtWithoutStructuralMatch visiting ty: {:?}", ty);
+
+            if let ty::Adt(adt_def, _) = ty.sty {
+                if !self.tcx.has_attr(adt_def.did, sym::structural_match) {
+                    self.found = Some(&adt_def);
+                    debug!("SearchForAdtWithoutStructuralMatch found adt_def: {:?}", adt_def);
+                    return true // Halt visiting
```
Comment (Centril) [inline src/librustc_mir/hair/pattern/mod.rs]:
```suggestion
                    return true // Halt visiting!
```

File: src/librustc_mir/hair/pattern/mod.rs [hunk 4]
```diff
@@ -1058,6 +1067,52 @@ impl<'a, 'tcx> PatternContext<'a, 'tcx> {
...
+
+                // For ADT, we do *not* visit contents.
+                //
+                // The presence of the #[structural_match] attribute
+                // certifies it for use in match.
+                //
+                // (Also, visiting contents would visit the
+                // substitutions of any type parameters on the ADT,
+                // which we do not want. Visitng them mishandles
+                // PhantomData<T> according to rust-lang/rust#55028
```
Comment (Centril) [inline src/librustc_mir/hair/pattern/mod.rs]:
```suggestion
                // `PhantomData<T>` according to rust-lang/rust#55028
```

File: src/test/ui/rfc1445/issue-61118-match-slice-forbidden-without-eq.rs [hunk 5]
```diff
@@ -0,0 +1,19 @@
+// Issue 61118 pointed out a case where we hit an ICE during code gen:
+// the compiler assumed that PartialEq was always implemented on any
```
Comment (Centril) [inline src/test/ui/rfc1445/issue-61118-match-slice-forbidden-without-eq.rs]:
```suggestion
// the compiler assumed that `PartialEq` was always implemented on any
```

File: src/test/ui/rfc1445/issue-61118-match-slice-forbidden-without-eq.rs [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+// const used in a pattern context, but the code responsible for
```
Comment (Centril) [inline src/test/ui/rfc1445/issue-61118-match-slice-forbidden-without-eq.rs]:
```suggestion
// `const` used in a pattern context, but the code responsible for
```

File: src/test/ui/rfc1445/issue-62307-match-ref-ref-forbidden-without-eq.rs [hunk 7]
```diff
@@ -0,0 +1,39 @@
+// RFC 1445 introduced `#[structural_match]`; this attribute must
+// appear on the struct/enum definition for any const used in a
```
Comment (Centril) [inline src/test/ui/rfc1445/issue-62307-match-ref-ref-forbidden-without-eq.rs]:
```suggestion
// appear on the `struct`/`enum` definition for any `const` used in a
```

File: src/test/ui/rfc1445/issue-62307-match-ref-ref-forbidden-without-eq.rs [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+// resolve the question of what semantics is used for such matching.
+// (See RFC 1445 for more details and discussion.)
+
+// Issue 62307 pointed out a case where the checking for
+// `#[structural_match]` was too shallow.
+
+#[derive(Debug)]
+struct B(i32);
+
+// Overriding PartialEq to use this strange notion of "equality" exposes
```
Comment (Centril) [inline src/test/ui/rfc1445/issue-62307-match-ref-ref-forbidden-without-eq.rs]:
```suggestion
// Overriding `PartialEq` to use this strange notion of "equality" exposes
```

File: src/test/ui/rfc1445/match-nonempty-array-forbidden-without-eq.rs [hunk 9]
```diff
@@ -0,0 +1,19 @@
+// Issue 62307 pointed out a case where the checking for
+// `#[structural_match]` was too shallow.
+//
+// Here we check similar behavior for non-empty arrays of types that
+// do not derive Eq.
```
Comment (Centril) [inline src/test/ui/rfc1445/match-nonempty-array-forbidden-without-eq.rs]:
```suggestion
// do not derive `Eq`.
```

**Review by Centril — COMMENTED** [review]
Just some nits...

**Comment by matthewjasper** [comment]
Does this handle matching on `const A: &[Option<NoDerive>] = ...`?

File: src/test/ui/rfc1445/issue-61118-match-slice-forbidden-without-eq.rs (hunk 6 above)
Comment (pnkfelix) [reply src/test/ui/rfc1445/issue-61118-match-slice-forbidden-without-eq.rs]: mmm, I usually don't put something in `code format` when I'm not using the syntactic form and am instead using the word to describe a mechanism or broader pattern. Sort of like how one would not put "trait" in `code format` unless one were talking about a bit of code that actually used the `trait` keyword.

Comment (Centril) [reply src/test/ui/rfc1445/issue-61118-match-slice-forbidden-without-eq.rs]: Fair :)

**Comment by pnkfelix** [comment]
@matthewjasper no, it mishandles the case of `const A: &[Option<NoDerive>]` (as in, it fails to reject a pattern using that const).

The code also mishandles some other cases that I just discovered.
 * I wanted to double check the text of rust-lang/rfcs#1445 to make sure I understand the intended scope of that change
 * its possible the RFC *intended* for us to use a shallow check of the attribute; i.e., that we would accept the case of a struct wrapping another struct, where the wrapper derives `PartialEq` and `Eq` but the wrapped type does not derive them and instead uses its own implementation.
 * But I don't think that could have been the true intention, since allowing such side-stepping of the attribute via a wrapper struct seems like it would undermine the objectives of the RFC itself.
 * (I *suspect* the reality is the RFC was written assuming a sort of recursive semantics, where converting the wrapper to a pattern would involve converting the wrapped struct to a pattern, which in turn would cause the wrapped struct to be checked. But this is not what currently happens, due to the way we sometimes dispatch to a `PartialEq::eq` call rather than expanding.)

I'll post a follow-up comment outlining the behavior on a set of cases after I finishing going through the analysis.

File: src/test/ui/rfc1445/issue-61118-match-slice-forbidden-without-eq.rs (hunk 6 above)
Comment (pnkfelix) [reply src/test/ui/rfc1445/issue-61118-match-slice-forbidden-without-eq.rs]: Would you prefer I use another word here? I did consider saying "constant" but then I worried that would then stray into the territory of describing a syntactic literal, and I need the reader to keep in mind that we are talking about uses of `const` items.

Maybe I should say it that way, "... that `PartialEq` was always implemented on any use of a `const` item in a pattern context, but ..." ?

**Comment by pnkfelix** [comment]
Okay, that last change (to move the visit up above the recursive loop) broke cases like these:

```rust
struct NoDerive(i32);

impl PartialEq for NoDerive {
    fn eq(&self, _: &Self) -> bool { false }
}

impl Eq for NoDerive { }

#[derive(PartialEq, Eq)]
struct WrapInline(NoDerive);

const WRAP_DIRECT_INLINE: WrapInline = WrapInline(NoDerive(0));

#[derive(PartialEq, Eq)]
struct WrapParam<T>(T);

const WRAP_DIRECT_PARAM: WrapParam<NoDerive> = WrapParam(NoDerive(0));
```

Meanwhile, regardless of where the visit is done, I also fail to correctly handle:

```rust
const WRAP_INDIRECT_INLINE: & &WrapInline = & &WrapInline(NoDerive(0));

const WRAP_INDIRECT_PARAM: & &WrapParam<NoDerive> = & &WrapParam(NoDerive(0));
```

**Comment by pnkfelix** [comment]
Okay so **now** I've removed commit 9db458f300a1f0c844071a41a547bd6294c99610 to avoid regressing cases that we *were* handling correctly before, and I've added 1e55df7a3ab2fca0ec35b884cf556af4b76d3624 because I've decided that we do need to visit the substs of an ADT (and just treat `PhantomData<T>` as a special case).

With that change in place, we handle @matthewjasper 's example correctly, as well as most of the other ones I identified.

The one case that this version continues to mishandle is this ([play](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=9fb46f1f29d393cc7bdf4568e07b7d65)):

```rust
struct NoDerive;

impl PartialEq for NoDerive {
    fn eq(&self, _: &Self) -> bool { false }
}

impl Eq for NoDerive { }

#[derive(PartialEq, Eq)]
struct WrapInline(NoDerive);

const WRAP_INLINE: & &WrapInline = & &WrapInline(NoDerive);

fn main() {
    match WRAP_INLINE {
        WRAP_INLINE => { println!("WRAP_INLINE matched WRAP_INLINE"); }
        _ => { println!("WRAP_INLINE did not match WRAP_INLINE"); }
    }
}
```

I think handling this case correctly (and also efficiently) will require deeper changes, though. E.g. a new bit in `rustc::ty::AdtFlags`.

 * To be honest we might be better off trying to address finding our long-term solution to the structural-match problem rather than trying to plug all the leaks in `#[structural_match]`...
 * But under the assumption that gathering consensus on a language problem like that will always take more time than you expect, I'll see about changing `rustc::ty::AdtFlags` in the short term.

File: src/test/ui/rfc1445/issue-61118-match-slice-forbidden-without-eq.rs (hunk 6 above)
Comment (Centril) [reply src/test/ui/rfc1445/issue-61118-match-slice-forbidden-without-eq.rs]: I wouldn't overthink it -- it was mostly a nit on my part. ;)
That said, your last suggestion makes sense.

**Comment by pnkfelix** [comment]
Okay, so *now* I've added a more complete traversal that addresses the problem posted above.

But doing that then led me to ask some questions, like: 
 * "Now that I am traversing the fields, should I stop traversing the substs of an ADT?"
   * (probably yes)
 * "How should unsafe pointers be handled?" 
   * (probably should not recur on the `T` in `*const T`)

The fact that I'm asking these questions leads me to think three things:
  * I need to add more tests
  * The implementation here needs more work
  * This probably needs to be initially deployed with a warning cycle. (I don't think that will be too hard to do, even.)

**Comment by pnkfelix** [comment]
Turning this from an error to a warning exposes *more* issues; e.g. the point where we are doing this check in the compiler's control flow, namely lowering from HIR to HAIR, is a spot where for some reason we may attempt to lower the same input multiple times, and thus the warning gets issued redundantly.

(The very fact that we are re-running lower on the same input seems like a worrisome thing, just in terms of compiler efficiency...)

So, do I focus on making this a `LintPass` rather than doing it in lowering? Or do I try to figure out why lowering is being run multiple times on the same input?

**Comment by pnkfelix** [comment]
Okay I'm finally narrowing in on a solution that I'm happy with, I hope.

I am going to try to deploy this as a future-compat lint; I've filed #62411 as a tracking issue for that.

**Comment by pnkfelix** [comment]
(last commit 02af3ca5c42f35eeceb0761180193a6abb97c0ad was result of my realization that while I had added [special case code](https://github.com/rust-lang/rust/pull/62339/commits/b56080162bc45e1cfd561b0a21e50b827c9df679#diff-e826542d6f24137d5b0b5ae333230529R1185) to preserve existing behavor for empty-arrays (#62336), I hadn't put in a corresponding unit test. That's fixed now.)

File: src/librustc_mir/hair/pattern/mod.rs [hunk 10]
```diff
@@ -1058,6 +1145,96 @@ impl<'a, 'tcx> PatternContext<'a, 'tcx> {
     }
 }
 
+fn search_for_adt_without_structural_match<'tcx>(tcx: TyCtxt<'tcx>,
+                                                 ty: Ty<'tcx>)
+                                                 -> Option<&'tcx AdtDef>
```
Comment (nikomatsakis) [inline src/librustc_mir/hair/pattern/mod.rs]: I think this function could use a comment giving more context and perhaps pointing to some of the relevant issues. It's name is good but would be pretty opaque if you didn't understand what was going on.

**Comment by nikomatsakis** [comment]
My 2 cents:

- r=me on the PR itself.
- it seems like a good idea to do a crater run (with this lint level set to deny) to get an idea of the impact.
- I do think we ought to resolve this question around constants. This feels like something where, if we had a functional working group around const generics, they could ultimately resolve it.

**Comment by pnkfelix** [comment]
hmm, looking at #36891, I now wonder: should we use a different name than `non_structural_match` for the future-compat lint here? I had considered e.g. `indirect_non_structural_match`...

Having said that, I nonetheless plan to let this PR land as is. We can bikeshed the names while the PR is on nightly (right?)

File: src/librustc_mir/hair/pattern/mod.rs (hunk 10 above)
Comment (pnkfelix) [reply src/librustc_mir/hair/pattern/mod.rs]: Good idea, added such a comment in b0b64dd

**Comment by pnkfelix** [comment]
(actually on second thought I probably will do some rename of the lint before landing it. the current name is a bit too broad for my taste...)

**Pull request merged**